servers/zms/src/main/java/com/yahoo/athenz/zms/DBService.java

```diff
@@ -7996,6 +8039,147 @@ public DomainList listDomainDependencies(String service) {
         }
     }
 
+    void executeDeleteExpiredMembership(ResourceContext ctx, ObjectStoreConnection con, String domainName, String roleName,
+                                        String normalizedMember, String auditRef, String caller) {
+
+        final String principal = getPrincipalName(ctx);
+
+        // process our delete role member operation
+
+        if (!con.deleteRoleMember(domainName, roleName, normalizedMember, principal, auditRef)) {
+            throw ZMSUtils.notFoundError(caller + ": unable to delete role member: " +
+                    normalizedMember + " from role: " + roleName, caller);
+        }
+    }
+
+    public void executeDeleteAllExpiredRoleMemberships(ResourceContext ctx, String auditRef, String caller) {
```

Comment:
we should somewhat careful about having all this in a single transaction. if our set contains users from 100 different domains, are we effectively blocking changes to those 100 domains until we commit our changes? How about if we fetch the list of users that need to be expunged, separate them per domain and then process each domain at a time. This way only a single domain is blocked and not all of them

Reply:
This could lead to an edge case where someone will update the expiration in the meantime.
Do we need to consider this scenario?

Reply:
yes, we should. if we generate our delete sql commands such that to include the expiry time in the where clause when then we'll never delete a member incorrectly where the expiry was changed after we read the record.